### ISSUE TITLE
Automatically mark mail is "read"

### ISSUE THREAD

**Issue by romanofski** [purebred-mua/purebred#53]
Now with write support in hs-notmuch, adding a way to reflect that mail has been viewed is possible. We could just easily remove the "unread" tag, write the rest of the tags to notmuch and be done with it when you view the mail.
Also add a keybinding which sets the mail back to "unread".

**Comment by romanofski** [comment]
I actually started with this on the side ...

**Comment by romanofski** [comment]
Got it together so that it builds (the lowest denominator) thanks to your example. A few outstanding things:

* The NotmuchMail record has a Boolean field I use for displaying to decide if it's new or not. I'll try to remove this and simply do this tasks during rendering. Because, it'll save me setting this value again after rewriting the tags.
* `mkTag` gives me a `Tag` of type Maybe. What I need tho is to convert all text tags to a Maybe [Tag] to signal that an error happened when writing/converting the tags. Not sure yet how I accomplish that.

**Comment by frasertweedale** [comment]
On Fri, Sep 01, 2017 at 05:06:25AM -0700, Roman Joost wrote:
> * `mkTag` gives me a `Tag` of type Maybe. What I need tho is to convert all text tags to a Maybe [Tag] to signal that an error happened when writing/converting the tags. Not sure yet how I accomplish that.

traverse :: (Traversable t, Applicative f) => (a -> f b) -> t a -> f (t b)

traverse ::                               (a -> Maybe b) -> [a] -> Maybe [b]

traverse mkTag (tags :: [Text])

**Comment by romanofski** [comment]
Hm... wondering about two things:

1. Instead of storing decoded `Text` as tags, maybe the tags should be decoded "just in time" for rendering? Would save me the back and forth of encoding/decoding when manipulating the tags.
1. I'm following the use case, that you nominate a tag reflecting an 'unread' state of mail and once you view the mail, that tag is removed. Thus the mail becoming "read". You can obviously toggle it back. Other mail clients follow a way to "write" only when you press a key combination. Which means you accumulate a set of commands and you explicitly "flush" them to disk. Thinking about it, for just reflecting a viewed state for Mail I would think this become a burden, since you will always need to flush those trivial changes. Any thoughts?

**Comment by romanofski** [comment]
@frasertweedale for `traverse` doesn't that mean, that you could end up with a shorter list of "parseable" tags?

**Comment by frasertweedale** [comment]
@romanofski no, for traverse; if any fail, they all fail.
```
λ> traverse id [Just 1, Just 2]
Just [1,2]
λ> traverse id [Nothing, Just 2]
Nothing
```

**Comment by frasertweedale** [comment]
> Hm... wondering about two things:
> 
> 1. Instead of storing decoded Text as tags, maybe the tags should be decoded "just in time" for rendering? Would save me the back and forth of encoding/decoding when manipulating the tags.

`Text` is semantically correct (although the Notmuch binding currently deals with `ByteString`, tags are supposed to be utf8-encoded strings**).  Also Haskell is non-strict.  I wouldn't worry about it just
now, and possibly not ever unless we have measurements to show that it's a problem.

** FWIW I might change *hs-notmuch* to use Text, but am undecided at this point.

> 2. I'm following the use case, that you nominate a tag reflecting an 'unread' state of mail and once you view the mail, that tag is removed. Thus the mail becoming "read". You can obviously toggle it back. Other mail clients follow a way to "write" only when you press a key combination. Which means you accumulate a set of commands and you explicitly "flush" them to disk. Thinking about it, for just reflecting a viewed state for Mail I would think this become a burden, since you will always need to flush those trivial changes. Any thoughts?

In the simple case, immediate write is fine IMO.  But we have a ticket for "marking" mail and performing bulk actions anyway, so when we get to that point we can review how to do this.

**Comment by frasertweedale** [comment]
Fixed in #56.
